### ISSUE TITLE
Exported Resources don't work

### ISSUE THREAD

**Issue by rychannel** [voxpupuli/puppet-icinga2#232]
On my Icinga server I get the following error when trying to pull in the exported resources.

Error: Could not retrieve catalog from remote server: Error 500 on SERVER: {"message":"Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, prefix(): expected first argument to be an Array or a Hash, got \"\" at /etc/puppetlabs/code/environments/test/modules/icinga2/manifests/config.pp:20:16 at /etc/puppetlabs/code/environments/test/modules/icinga2/manifests/object/host.pp:218 on node tst-ici-as-1.hoad.local","issue_kind":"RUNTIME_ERROR"}

My class for that server just includes

```
class roles::icinga::appserver {

  Icinga2::Object::Host <<||>>

}
```

**Comment by lbetz** [comment]
That's new. Please tell me your distribution and the puppet version you used. Did you notice the requirements? The required modules in the correct version especially the stdlib?

**Comment by lazyfrosch** [comment]
Wild guess: no include of `::icinga2` before realizing objects?

**Comment by snahelou** [comment]
Same problem for me when I use icinga2::feature::idopgsql it crash randomly

My setup: 

Work: icinga2 + icinga2::feature::command
Doesn't work: icinga2 + icinga2::feature::command + icinga2::feature::idopgsql

**Comment by rychannel** [comment]
Rhel 7 and puppet 4.8 (PE 2016.5.1)

**Comment by rychannel** [comment]
I have the correct version of Stdlib installed.

**Comment by bobapple** [comment]
@rychannel can you add some more puppet code so we can try to reproduce this?

**Comment by rychannel** [comment]
Here is the profile that runs on the clients (which are Windows 2012 R2 btw)

```
class profiles::icinga::hostcheck
{
  @@icinga2::object::host { "${::fqdn}":
    display_name  => "${::fqdn}",
    address       => "${::ipaddress}",
    check_command => 'hostalive',
    target        => "/etc/icinga2/objects/hosts/${::fqdn}.conf",
  }
}
```

**Comment by bobapple** [comment]
Does the `fqdn` fact work on your windows machine?

**Comment by rychannel** [comment]
yes

**Comment by bobapple** [comment]
Can you dump the resource from your PuppetDB?

For me it looks like one of your variables is empty, but its hard to tell without further information

**Comment by bobapple** [comment]
I'm closing this due to lack of information

**Comment by TinajaLabs** [comment]
Doh! - I'm getting the same error.

> Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Function Call, prefix(): expected first argument to be an Array or a Hash, got "" at /opt/puppetlabs/puppet/modules/icinga2/manifests/config.pp:20:16 on node mon.sys-f.xxx
> Warning: Not using cache on failed catalog
> Error: Could not retrieve catalog; skipping run

This seems to be the offending line in /manifests/config.pp:
```  $constants = prefix($::icinga2::_constants, 'const ')```

I thought it might be error of empty variable in my constants array, defined in class  '::icinga2': , but cannot identify any problems.

- Current version of puppet-icinga2
- rhel 6.7
- puppet agent 4.8.1
- puppet master 4.8.1

trying to learn how to set up master client zones with icinga as agent to replace nrpe

some sample code below ~ I have 2 sections, one for when I run puppet on master/mon it installs everything, on client it just installs icinga, checks, basic configs.  I have also tried commenting the entire constants sections - still the problem occurs.

I discovered that the config.pp is apparently called in config_spec.rb and fragmant_spec.rb but I'm not smart enough to know what is going on in these ruby files.

```
	# for MASTER -  master node needs to have hostname = 'mon'

	if "mon" in $hostname {

		class { '::icinga2':
			manage_package	=> true,
			confd 			=> true,
			features 		=> ['checker','mainlog','notification','statusdata','compatlog','command'],
			constants		=> {
				'ZoneName' 				=> 'zone_master',
				'NodeName' 				=> "${::fqdn}",
				'PluginDir'				=> '/etc/icinga2/libexec',
				'CustomPluginDir'		=> '/xxx/icinga/plugins',
				'PluginContirbDir' 		=> '/usr/local/nagios/plugins',
				'TrackName' 			=> "$track_name",
				'SourceName' 			=> 'mon.src.xxx',
				'TicketSalt' 			=> "$ticket_salt",
				'baseWikiURL' 			=> 'http://10.0.1.12/tw/index.php/',
				'baseSnmpWalk' 			=> 'snmpwalk -v 2c -c public ',
				'monlevel' 				=> '2',
			},
		}

		class { '::icinga2::feature::api':
			accept_commands	=> true,
			accept_config 	=> false,
			endpoints       => {
			'NodeName'   		=> {},
			"$master_fqdn"=> {
				'host' 	=> "$master_ip",
			},
			'db01.sys-f.xxx' 	=> {
				'host' 	=> '192.168.1.28',
			},
			'ui.sys-f.xxx' 	=> {
				'host' 	=> '192.168.1.23',
			},
			},
			zones           => {
			'zone_master' => {
				'endpoints' => ["$master_fqdn"],
			},
			"zone_${track_name}"    => {
				'endpoints' => ['db01.sys-f.xxx','ui.sys-f.xxx'],
				'parent'    => 'master',
			},
			},
		}
	} else {

		# for all CLIENTS

		class { '::icinga2':
			manage_package	=> true,
			confd 			=> false,
			features 		=> ['checker','mainlog'],
			constants		=> {
				'NodeName' 				=> "${::fqdn}",
				'PluginDir'				=> '/etc/icinga2/libexec',
				'CustomPluginDir'		=> '/xxx/icinga/plugins',
				'PluginContirbDir' 		=> '/usr/local/nagios/plugins',
				'TrackName' 			=> "$track_name",
				'SourceName' 			=> 'mon.src.xxx',
				'TicketSalt' 			=> "$ticket_salt",
				'baseWikiURL' 			=> 'http://10.0.1.12/tw/index.php/',
				'baseSnmpWalk' 			=> 'snmpwalk -v 2c -c public ',
				'monlevel' 				=> '2',
			},
		}

		# build the endpoints and zones for clients
		class { '::icinga2::feature::api':
		  pki             => 'none',
		  accept_commands => true,
		  accept_config   => true,
		  endpoints       => {
		    'NodeName'    => {},
		    "$master_fqdn" => {
		      'host' => "$master_ip",
		    }
		  },
		  zones           => {
		    'ZoneName' => {
		      'endpoints' => ['NodeName'],
		      'parent'    => 'zone_master',
		    },
		    'zone_master'      => {
		      'endpoints' => ["$master_fqdn"],
		    }
		  }
		}
	}

	# all hosts get this
	icinga2::object::zone { 'global-templates':
		global => true,
	}

Thanks for any tips,
Chris.

**Comment by lazyfrosch** [comment]
@TinajaLabs You are not even using exported resources. Your problem might be an ordering problem, where you use a feature or object, before including the main class icinga2.

See also #268 

Please make sure that you are not realizing / declaring any `icinga2::*` defined resources or classes before `::icinga2`

And please open a new issue :)

**Comment by TinajaLabs** [comment]
Apologies.  I did not see the issue #268 that you referenced.  I only noticed the error was the same.

I have now narrowed the issue to the same code related to ```idomysql```.

Further I see now that we deploy against redhat 6.7 and it does indeed default to 1.8.7.  RVM is not even in our repo.  We are not a ruby shop and just use it to make puppet happen.  Our deployments are many and require offline repos that we build manually.  I doubt we will switch to 1.9+ anytime soon.  I will see if I can bypass this function; do it brute force outside of puppet-icinga.  If not, this is tragic for me.  This seems the best module for puppeting icinga.

Thanks for your help, Chris.
